File: src/main/java/me/taubsie/dungeonhub/application/command/commands/UserScanCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.exceptions.MissingPermissionException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.application.service.ProfileModerationService;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.message.Message;
import org.javacord.api.entity.message.MessageFlag;
import org.javacord.api.entity.message.component.ActionRowBuilder;
import org.javacord.api.entity.message.component.ButtonBuilder;
import org.javacord.api.entity.message.component.ButtonStyle;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.server.Server;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.ButtonInteraction;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;
import org.javacord.api.interaction.callback.InteractionOriginalResponseUpdater;

import java.util.*;
import java.util.stream.Collectors;

@CommandParameters(name = "userscan",
        description = "Scans for users with a bad username.",
        enabledForPermissions = {PermissionType.BAN_MEMBERS})
public class UserScanCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        Server server = getServer();

        if (!server.hasPermission(slashCommandCreateEvent.getSlashCommandInteraction().getUser(),
                PermissionType.ADMINISTRATOR)) {
            throw new MissingPermissionException();
        }

        boolean ban = slashCommandCreateEvent.getSlashCommandInteraction().getOptionByName("ban").isPresent();

        Map<User, String> result = new HashMap<>();
        Map<User, String> excluded = new HashMap<>();

        InteractionOriginalResponseUpdater updater = slashCommandCreateEvent.getSlashCommandInteraction()
                .respondLater().join();

        for (User user : server.getMembers()) {
            if (!user.isBot()) {
                String checkResult = ProfileModerationService.getInstance().checkUserName(user.getName());
                if (checkResult != null) {
                    if (ProfileModerationService.getInstance().isExcluded(user, server)) {
                        excluded.put(user, checkResult);
                    } else {
                        result.put(user, checkResult);
                    }
                }
            }
        }

        if (ban) {
            for (Map.Entry<User, String> entries : result.entrySet()) {
                ProfileModerationService.getInstance().handleUserBan(server, entries.getKey(), entries.getValue());
            }
        }

        updater.addEmbed(ApplicationService.getInstance()
                        .getEmbed()
                        .setColor(EmbedColor.NEGATIVE.getColor())
                        .setDescription((ban ? "Banned" : "Flagged")
                                + ":\n" + result.entrySet()
                                .stream()
                                .map(userStringEntry ->
                                        userStringEntry.getKey().getMentionTag() + " | " + userStringEntry.getValue())
                                .collect(Collectors.joining("\n"))
                                + "\n\nExcluded:\n" +
                                excluded.entrySet()
                                        .stream()
                                        .map(userStringEntry ->
                                                userStringEntry.getKey().getMentionTag() + " | " + userStringEntry.getValue())
                                        .collect(Collectors.joining("\n"))));

                updater.addComponents(new ActionRowBuilder().addComponents(
                        new ButtonBuilder().setStyle(ButtonStyle.DANGER)
                                .setLabel("Show " + ((ban) ? "banned" : "flagged") + " users")
                                .setCustomId("show_flagged_banned")
                                .build(),
                        new ButtonBuilder().setStyle(ButtonStyle.SECONDARY)
                                .setLabel("Show excluded users")
                                .setCustomId("show_excluded")
                                .build()
                ).build());

        Message sentMessage = updater.update().join();

        sentMessage.addButtonClickListener(event -> {
            event.getButtonInteractionWithCustomId("show_flagged_banned").ifPresent(buttonInteraction -> respondWithUsers(buttonInteraction, result.keySet(), server));

            event.getButtonInteractionWithCustomId("show_excluded").ifPresent(buttonInteraction -> respondWithUsers(buttonInteraction, excluded.keySet(), server));
        });
    }

    //TODO test
    //TODO maybe implement more logic to this idk?
    private void respondWithUsers(ButtonInteraction buttonInteraction, Set<User> users, Server server) {
        String content = users.stream().map(DiscordEntity::getIdAsString).collect(Collectors.joining(","));

        EmbedBuilder embed = ApplicationService.getInstance()
                .getEmbed()
                .setDescription(users.stream().map(user -> user.getMentionTag() + " | " + user.getDisplayName(server)).collect(Collectors.joining("\n")));

        buttonInteraction.createImmediateResponder()
                .setContent(content)
                .addEmbed(embed)
                .setFlags(MessageFlag.EPHEMERAL)
                .respond();
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption banCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("ban")
                .setDescription("Add this if flagged users should also be banned.")
                .build();

        SlashCommandOption scanCommand = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("scan")
                .setDescription("Add this if flagged users shouldn't be banned.")
                .build();

        return Arrays.asList(banCommand, scanCommand);
    }
}

File: src/main/java/me/taubsie/dungeonhub/kord/application/commands/UserScanCommand.kt
package me.taubsie.dungeonhub.kord.application.commands

import com.kotlindiscord.kord.extensions.checks.hasPermission
import com.kotlindiscord.kord.extensions.commands.application.slash.publicSubCommand
import com.kotlindiscord.kord.extensions.components.components
import com.kotlindiscord.kord.extensions.components.ephemeralButton
import com.kotlindiscord.kord.extensions.extensions.Extension
import com.kotlindiscord.kord.extensions.extensions.publicSlashCommand
import dev.kord.common.entity.ButtonStyle
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kord.core.entity.Guild
import dev.kord.core.entity.Member
import dev.kord.core.entity.User
import dev.kord.core.entity.effectiveName
import dev.kord.rest.builder.message.EmbedBuilder
import dev.kord.rest.builder.message.create.FollowupMessageCreateBuilder
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import me.taubsie.dungeonhub.kord.application.enums.EmbedColor
import me.taubsie.dungeonhub.kord.application.loader.LoadExtension
import me.taubsie.dungeonhub.kord.application.service.ApplicationService
import me.taubsie.dungeonhub.kord.application.service.ProfileModerationService
import java.util.stream.Collectors

@LoadExtension
class UserScanCommand : Extension() {
    override val name = "user-scan-command"

    override suspend fun setup() {
        publicSlashCommand {
            name = "userscan"
            description = "Scans for users with a bad username."
            defaultMemberPermissions = Permissions(Permission.BanMembers)
            allowInDms = false

            check {
                hasPermission(Permission.Administrator)
            }

            publicSubCommand {
                name = "scan"
                description = "Add this if flagged users shouldn't be banned."

                action {
                    respond(execute(false, guild!!.asGuild()))
                }
            }

            publicSubCommand {
                name = "ban"
                description = "Add this if flagged users should also be banned."

                action {
                    respond(execute(true, guild!!.asGuild()))
                }
            }
        }
    }

    suspend fun execute(ban: Boolean, guild: Guild): FollowupMessageCreateBuilder.() -> Unit {
        val result: MutableMap<Member, String> = HashMap()
        val excluded: MutableMap<Member, String> = HashMap()

        for (member in guild.members.toList()) {
            if (!member.isBot) {
                val checkResult = ProfileModerationService.checkUserName(member.globalName)
                if (checkResult != null) {
                    if (ProfileModerationService.isExcluded(member, guild)) {
                        excluded[member] = checkResult
                    } else {
                        result[member] = checkResult
                    }
                }
            }
        }

        if (ban) {
            for ((key, value) in result) {
                ProfileModerationService.handleUserBan(guild, key, value)
            }
        }

        val embed = ApplicationService.embed
        embed.color = EmbedColor.NEGATIVE.color
        embed.description = ((if (ban) "Banned" else "Flagged")
                + ":\n" + result.entries
            .stream()
            .map { userStringEntry: Map.Entry<Member, String> -> userStringEntry.key.mention + " | " + userStringEntry.value }
            .collect(Collectors.joining("\n"))
                + "\n\nExcluded:\n" +
                excluded.entries
                    .stream()
                    .map { userStringEntry: Map.Entry<Member, String> -> userStringEntry.key.mention + " | " + userStringEntry.value }
                    .collect(Collectors.joining("\n")))

        return {
            embeds = mutableListOf(embed)

            runBlocking {
                launch {
                    components {
                        ephemeralButton {
                            style = ButtonStyle.Danger
                            id = "show_flagged_banned"
                            label = "Show " + (if ((ban)) "banned" else "flagged") + " users"

                            action {
                                respond {
                                    content = getContent(result.keys)
                                    embeds = mutableListOf(getEmbed(result.keys))
                                }
                            }
                        }

                        ephemeralButton {
                            style = ButtonStyle.Secondary
                            id = "show_excluded"
                            label = "Show excluded users"

                            action {
                                respond {
                                    content = getContent(excluded.keys)
                                    embeds = mutableListOf(getEmbed(excluded.keys))
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    private fun getContent(users: Set<User>): String {
        return users.stream().map { obj: User -> obj.id.toString() }.collect(Collectors.joining(","))
    }

    private fun getEmbed(users: Set<Member>): EmbedBuilder {
        val embed = ApplicationService.embed
        embed.description = users.stream().map { user: User ->
            user.mention + " | " + user.effectiveName
        }.collect(Collectors.joining("\n"))

        return embed
    }
}